src/linker/Linker/MessageContainer.cs

```diff
@@ -49,6 +49,7 @@ public static MessageContainer CreateErrorMessage (string text, int code, string
 			if (!(code >= 1000 && code <= 2000))
 				throw new ArgumentException ($"The provided code '{code}' does not fall into the error category, which is in the range of 1000 to 2000 (inclusive).");
 
+			origin?.TryGetSourceInfo ();
```

Comment:
This is an odd code flow, you are changing the content of struct from a constructor of container-like object. The logic is relevant to MessageOrigin::ToString only and I think it'd be better to do it there only. This would allow change MessageOrigin to be readonly and ensure that logic in methods like MessageOrigin::Equals work all the time

Reply:
I agree with @marek-safar here. Currently we only use the info from `TryGetSource` in two places - in the warning case below where we prepend the location from member if there's no file name and then in the ToString.
I think we should unify this:
* Same behavior between error and warning around "origin/location"
* Remove `TryGetSourceInfo` and make the struct `readonly`
* In `ToString` if we have `FileName` keep the current behavior. If we don't have `FileName` use the "member" to describe the location.

Note: I think it's OK for the linker to report warnings which don't have the typical `filename(line)` information since we're dealing with IL input. That said in most cases we know where the error/warning comes from, so we should report it. Unfortunately we have to "invent" our own syntax for this, but that's also OK.